### PR TITLE
ref: type sentry/utils/assets.py and sentry/utils/http.py

### PR DESCRIPTION
**Pull request by asottile-sentry**

### CODE REVIEW

File: src/sentry/integrations/bitbucket/client.py [hunk 1]
```diff
@@ -70,7 +73,7 @@ def get_issues(self, repo):
     def create_issue(self, repo, data):
         return self.post(path=BitbucketAPIPath.issues.format(repo=repo), data=data)
 
-    def search_issues(self, repo, query):
+    def search_issues(self, repo: str, query: str) -> dict[str, Any]:
```
Comment (armenzg) [inline src/sentry/integrations/bitbucket/client.py:76]: Should it be `str` twice?
```suggestion
    def search_issues(self, repo: str, query: str) -> dict[str, str]:
```

File: src/sentry/integrations/bitbucket_server/integration.py [hunk 2]
```diff
@@ -260,8 +260,7 @@ def get_repositories(self, query=None):
                 for repo in resp.get("values", [])
             ]
 
-        full_query = (query).encode("utf-8")
-        resp = self.get_client().search_repositories(full_query)
+        resp = self.get_client().search_repositories(query)
```
Comment (armenzg) [inline src/sentry/integrations/bitbucket_server/integration.py:263]: Is this because before 3.8 we were not encoding automatically to utf?

File: src/sentry/integrations/utils/atlassian_connect.py [hunk 3]
```diff
@@ -28,10 +30,10 @@ def get_query_hash(
     for k, v in sorted(query_params.items()):
         # don't include jwt query param
         if k != "jwt":
-            if isinstance(v, list):
```
Comment (armenzg) [inline src/sentry/integrations/utils/atlassian_connect.py:31]: I'm seeing an inversion of the conditionals but nothing else. What's the reason for the change? Is checking for a string first better in some way?

File: src/sentry/integrations/utils/atlassian_connect.py [hunk 4]
```diff
@@ -88,7 +90,7 @@ def get_integration_from_jwt(
 
 
 def verify_claims(
-    claims: Optional[Mapping[str, str]],
+    claims: Mapping[str, str],
```
Comment (armenzg) [inline src/sentry/integrations/utils/atlassian_connect.py:93]: How did you noticed that this does not have to be optional anymore?

File: src/sentry/utils/http.py [hunk 5]
```diff
@@ -1,15 +1,24 @@
...
+from typing import TYPE_CHECKING, NamedTuple, Optional
 from urllib.parse import quote, urljoin, urlparse
 
 from django.conf import settings
+from django.http import HttpRequest
 
 from sentry import options
 
-ParsedUriMatch = namedtuple("ParsedUriMatch", ["scheme", "domain", "path"])
+if TYPE_CHECKING:
```
Comment (armenzg) [inline src/sentry/utils/http.py:11]: Why is this necessary?

File: src/sentry/utils/http.py [hunk 6]
```diff
@@ -20,14 +29,14 @@ def absolute_uri(url: Optional[str] = None, url_prefix=None) -> str:
     return urljoin(url_prefix.rstrip("/") + "/", url.lstrip("/"))
 
 
-def origin_from_url(url):
+def origin_from_url(url: str) -> str:
     if not url:
         return url
-    url = urlparse(url)
-    return f"{url.scheme}://{url.netloc}"
+    parsed = urlparse(url)
```
Comment (armenzg) [inline src/sentry/utils/http.py:35]: Good catch! Do not touch those input variables! :)

File: src/sentry/utils/http.py [hunk 7]
```diff
@@ -133,7 +133,7 @@ def is_valid_origin(origin, project=None, allowed=None):
             parsed_hostname = parsed.hostname
 
     if parsed.port:
-        domain_matches = (
+        domain_matches: tuple[str, ...] = (
```
Comment (armenzg) [inline src/sentry/utils/http.py:136]: How is this type understood? We have a tuple of at least one string?

If I wanted a tuple of at least two strings would I do `tuple[str, str, ...]`?

File: src/sentry/utils/http.py [hunk 8]
```diff
@@ -190,8 +190,6 @@ def origin_from_request(request):
     return rv
 
 
-def percent_encode(val):
+def percent_encode(val: str) -> str:
     # see https://en.wikipedia.org/wiki/Percent-encoding
-    if isinstance(val, str):
```
Comment (armenzg) [inline src/sentry/utils/http.py:195]: 🎉

File: src/sentry/integrations/bitbucket/client.py (hunk 1 above)
Comment (asottile-sentry) [reply src/sentry/integrations/bitbucket/client.py:76]: this returns a json blob -- I don't know the shape of it unfortunately but it's not `dict[str, str]`

File: src/sentry/integrations/bitbucket_server/integration.py (hunk 2 above)
Comment (asottile-sentry) [reply src/sentry/integrations/bitbucket_server/integration.py:263]: the `.encode("utf-8")` was a python-2-ism here -- the urlencoding functionalities understand both bytes and text in python 3

File: src/sentry/integrations/utils/atlassian_connect.py (hunk 3 above)
Comment (asottile-sentry) [reply src/sentry/integrations/utils/atlassian_connect.py:31]: the type annotation is `str | Sequence[str]` and a narrowing via `isinstance(..., list)` doesn't actually narrow anything because not all `Sequence`s are `list`s

File: src/sentry/integrations/utils/atlassian_connect.py (hunk 4 above)
Comment (asottile-sentry) [reply src/sentry/integrations/utils/atlassian_connect.py:93]: I traced the output of the `jwt` methods and noticed they only raise or return a dict

File: src/sentry/utils/http.py (hunk 5 above)
Comment (asottile-sentry) [reply src/sentry/utils/http.py:11]: there's an import cycle :'( -- the `*` imports in `sentry.models.__init__` suck :(

File: src/sentry/utils/http.py (hunk 6 above)
Comment (asottile-sentry) [reply src/sentry/utils/http.py:35]: heh yeah mypy screamed!

File: src/sentry/utils/http.py [hunk 9]
```diff
@@ -112,15 +121,6 @@ def is_valid_origin(origin, project=None, allowed=None):
     if origin == "null":
         return False
 
-    if isinstance(origin, bytes):
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/utils/http.py:115]: yep yep -- mypy reports "there is no valid class that is a subclass of both str and bytes" or something like that

**Review by armenzg — APPROVED** [review]

File: src/sentry/utils/http.py (hunk 7 above)
Comment (asottile-sentry) [reply src/sentry/utils/http.py:136]: I go over that in https://youtu.be/yLwvOwTO068 -- this is a variable-length tuple of strings

to represent "at least two strings" you need new functionality:

```python
# using compatible syntax
x: tuple[tuple[str], Unpack[tuple[str, ...]]]
# using the 3.11+ syntax
x: tuple[str, *tuple[str, ...]]
```

File: src/sentry/utils/jwt.py [hunk 10]
```diff
@@ -49,7 +50,7 @@ def decode(
     *,  # Force passing optional arguments by keyword
     audience: Optional[Union[str, bool]] = None,
     algorithms: Optional[List[str]] = None,
-) -> JSONData:
+) -> dict[str, JSONData]:
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/utils/jwt.py:53]: it was mad about "returning Any" since `JSONData` is an alias to `Any` -- I just pushed the `Any` one level deeper (and matched the annotation from `jwt`)

File: src/sentry/integrations/jira/client.py [hunk 11]
```diff
@@ -127,13 +127,13 @@ def update_comment(self, issue_key, comment_id, comment):
     def get_projects_list(self):
         return self.get_cached(self.PROJECT_URL)
 
-    def get_project_key_for_id(self, project_id):
+    def get_project_key_for_id(self, project_id) -> str:
```
Comment (evanh) [inline src/sentry/integrations/jira/client.py:130]: Shouldn't `project_id` have a type?

Comment (asottile-sentry) [reply src/sentry/integrations/jira/client.py:130]: it should, but this isn't being type checked yet and I don't know what the type is

Comment (evanh) [reply src/sentry/integrations/jira/client.py:130]: 👍

File: src/sentry/integrations/bitbucket/integration.py [hunk 12]
```diff
@@ -95,8 +95,8 @@ def get_repositories(self, query=None):
                 for repo in resp.get("values", [])
             ]
 
-        exact_query = f'name="{query}"'.encode()
-        fuzzy_query = f'name~"{query}"'.encode()
+        exact_query = f'name="{query}"'
```
Comment (untitaker) [inline src/sentry/integrations/bitbucket/integration.py:98]: you're changing the type of this variable, are we sure that is safe and not just broken typing further down the road? assuming the code currently works fine

File: src/sentry/utils/http.py [hunk 13]
```diff
@@ -120,15 +129,6 @@ def is_valid_origin(origin, project=None, allowed=None):
...
```
Comment (untitaker) [inline src/sentry/utils/http.py:123]: I am not convinced this is safe. What if this function is called from untyped code?

**Review by untitaker — COMMENTED** [review]
i'm not sure i would trust existing type annotations to cover the range of valid input types for a function.

File: src/sentry/integrations/bitbucket/integration.py (hunk 12 above)
Comment (asottile-sentry) [reply src/sentry/integrations/bitbucket/integration.py:98]: yes that's intentional and part of the change -- I surveyed all callsites and the `bytes` here was incorrect (it was being coerced to bytes and then out of bytes for no reason) -- these are also fairly well covered by tests

File: src/sentry/utils/http.py [hunk 14]
```diff
@@ -1,16 +1,25 @@
...
+
+
+class ParsedUriMatch(NamedTuple):
+    scheme: str
+    domain: str
+    path: str
 
 
-def absolute_uri(url: Optional[str] = None, url_prefix=None) -> str:
+def absolute_uri(url: Optional[str] = None, url_prefix: str | None = None) -> str:
```
Comment (ethanhs) [inline src/sentry/utils/http.py:22]:
```suggestion
def absolute_uri(url: str | None = None, url_prefix: str | None = None) -> str:
```
